File: tree/Traversal/Postorder/iterative/cpp/PostorderTraversal.cpp
#include "PostorderTraversal.hpp"

#include <utility>
#include <vector>

using namespace std;

vector<int> PostorderTraversal(const vector<int>& leftChildren, const vector<int>& rightChildren, int root) {
    vector<int> order;
    vector<pair<int, bool>> stack = {{root, false}};

    while (!stack.empty()) {
        const auto [node, expanded] = stack.back();
        stack.pop_back();

        if (expanded) {
            order.push_back(node);
            continue;
        }

        stack.push_back({node, true});

        if (rightChildren[node] != -1) {
            stack.push_back({rightChildren[node], false});
        }

        if (leftChildren[node] != -1) {
            stack.push_back({leftChildren[node], false});
        }
    }

    return order;
}

File: tree/Traversal/Postorder/iterative/rust/src/PostorderTraversal.rs
#[allow(non_snake_case)]
pub fn PostorderTraversal(left_children: &[i32], right_children: &[i32], root: usize) -> Vec<usize> {
    let mut order = Vec::new();
    let mut stack = vec![(root, false)];

    while let Some((node, expanded)) = stack.pop() {
        if expanded {
            order.push(node);
            continue;
        }

        stack.push((node, true));

        let right_child = right_children[node];
        if right_child != -1 {
            stack.push((right_child as usize, false));
        }

        let left_child = left_children[node];
        if left_child != -1 {
            stack.push((left_child as usize, false));
        }
    }

    order
}
